### PR TITLE
Enforce proper line ending removal when reading alphabet

### LINKED ISSUE #2611: CRLF in alphabet file breaks usage
```
python ./DeepSpeech.py --train_files ../zh-cn/clips/train.csv --dev_files ../zh-cn/clips/dev.csv --test_files ../zh-cn/clips/test.csv

Traceback (most recent call last):
  File "/mnt/d/FPProject_git/My_Work/speech/mozilla/DeepSpeech/util/text.py", line 33, in _label_from_string
    return self._str_to_label[string]
KeyError: '母'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/mnt/d/FPProject_git/My_Work/speech/mozilla/DeepSpeech/util/text.py", line 130, in text_to_char_array
    transcript = np.asarray(alphabet.encode(series['transcript']))
  File "/mnt/d/FPProject_git/My_Work/speech/mozilla/DeepSpeech/util/text.py", line 47, in encode
    res.append(self._label_from_string(char))
  File "/mnt/d/FPProject_git/My_Work/speech/mozilla/DeepSpeech/util/text.py", line 39, in _label_from_string
    ).with_traceback(e.__traceback__)
  File "/mnt/d/FPProject_git/My_Work/speech/mozilla/DeepSpeech/util/text.py", line 33, in _label_from_string
    return self._str_to_label[string]
KeyError: "ERROR: Your transcripts contain characters (e.g. '母') which do not occur in data/alphabet.txt! Use util/check_characters.py to see what characters are in your [train,dev,test].csv transcripts, and then add all these to data/alphabet.txt."

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "./DeepSpeech.py", line 965, in <module>
    absl.app.run(main)
  File "/home/kms/anaconda3/envs/speechenv/lib/python3.6/site-packages/absl/app.py", line 299, in run
    _run_main(main, args)
  File "/home/kms/anaconda3/envs/speechenv/lib/python3.6/site-packages/absl/app.py", line 250, in _run_main
    sys.exit(main(argv))
  File "./DeepSpeech.py", line 938, in main
    train()
  File "./DeepSpeech.py", line 438, in train
    train_phase=True)
  File "/mnt/d/FPProject_git/My_Work/speech/mozilla/DeepSpeech/util/feeding.py", line 101, in create_dataset
    df['transcript'] = df.apply(text_to_char_array, alphabet=Config.alphabet, result_type='reduce', axis=1)
  File "/home/kms/anaconda3/envs/speechenv/lib/python3.6/site-packages/pandas/core/frame.py", line 6928, in apply           return op.get_result()
  File "/home/kms/anaconda3/envs/speechenv/lib/python3.6/site-packages/pandas/core/apply.py", line 186, in get_result       return self.apply_standard()
  File "/home/kms/anaconda3/envs/speechenv/lib/python3.6/site-packages/pandas/core/apply.py", line 292, in apply_standard
    self.apply_series_generator()
  File "/home/kms/anaconda3/envs/speechenv/lib/python3.6/site-packages/pandas/core/apply.py", line 321, in apply_series_generator
    results[i] = self.f(v)
  File "/home/kms/anaconda3/envs/speechenv/lib/python3.6/site-packages/pandas/core/apply.py", line 112, in f
    return func(x, *args, **kwds)
  File "/mnt/d/FPProject_git/My_Work/speech/mozilla/DeepSpeech/util/text.py", line 136, in text_to_char_array
    raise ValueError('While processing: {}\n{}'.format(series['wav_filename'], e))
ValueError: ('While processing: /mnt/d/FPProject_git/My_Work/speech/mozilla/zh-cn/clips/common_voice_zh-CN_18782225.wav\n"ERROR: Your transcripts contain characters (e.g. \'母\') which do not occur in data/alphabet.txt! Use util/check_characters.py to see what characters are in your [train,dev,test].csv transcripts, and then add all these to data/alphabet.txt."', 'occurred at index 550')
```

I am training deep speech model in chinese.
I have already downloaded chiense dataset from voice.mozilla.org (Common Voice).
Then, I tried training as the description in Project Document(https://github.com/mozilla/DeepSpeech/blob/master/TRAINING.rst#training-your-own-model).
But when I train data using DeepSpeech.py, it failed with above errors.
data/alphabet.txt already contains '母'.
I can't find reason.
Please help me.

### PR DESCRIPTION
**Pull request by lissyx**
Fixes #2611

### CODE REVIEW

**Comment by lissyx** [comment]
> Finally, maybe a cleaner fix instead of removing the endings with str.replace would be using `open(..., encoding='utf8')` instead of `codecs.open`, which does newline normalization.

I was unsure about that, the combination of `codecs.open` + manual `[:-1]` I thought you had somethign specific in mind :)



> Can't you use the public encode/decode API to test this? To avoid touching private methods.

On the Alphabet object? Yeah, it should work.



> Also, can't we check in actual test files into the tree to avoid the mocking magic?

I actually think that for unit tests, it's better not to. This way, people can run them locally and be assured it won't shit in their desktop :)

**Comment by reuben** [comment]
> I was unsure about that, the combination of codecs.open + manual [:-1] I thought you had somethign specific in mind :)

Just an oversight. And maybe also related to Python 2 + 3 support. I don't remember.

> I actually think that for unit tests, it's better not to. This way, people can run them locally and be assured it won't shit in their desktop :)

As long as we check in the files they should still have the correct line endings on all platforms, no? And by overriding the behavior we risk testing something that's not actually what Python is doing. I don't know, I'm just deeply suspicious of mocking after seeing how badly it can fail in Gaia.

**Comment by lissyx** [comment]
> > 
> 
> As long as we check in the files they should still have the correct line endings on all platforms, no?

Ok, I better understand what you mean here. Why not, then.

**Comment by kdavis-mozilla** [comment]
I guess I should wait on the review until you've added in Reuben's suggestions

**Comment by lissyx** [comment]
> I guess I should wait on the review until you've added in Reuben's suggestions

You can go ahead, I've implemented those :)

File: util/test_text.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+from .text import Alphabet
+
+class TestAlphabetParsing(unittest.TestCase):
+
+    def _ending_tester(self, file, expected):
+        alphabet = Alphabet(os.path.join(os.path.dirname(__file__), 'test_data', file))
+        label = ''
+        for expected_label, expected_id in expected:
+            try:
+                label = alphabet.encode(expected_label)
```
[the remark this answers is not in the document]
Comment (lissyx) [reply util/test_text.py]: @kdavis-mozilla should be addressed

**Pull request merged**